### PR TITLE
Kafka input adapter

### PR DESCRIPTION
**Pull request by ryzhyk**
Simple adapter that reads data from Kafka into DBSP and doesn't do
anything clever for transactions or reliable delivery.  It is suitable
for streaming data in scenarios where atomic transactions are not
required and at least once delivery is acceptable (i.e., occasional
duplicates can be tolerated).

### CODE REVIEW

**Comment by mbudiu-vmw** [comment]
How is this tested? Do you need to install kafka on your machine?

**Comment by ryzhyk** [comment]
> How is this tested? Do you need to install kafka on your machine?

Good point.  I'll add a CI action to install redpanda.

File: adapters/src/test/data.rs [hunk 1]
```diff
@@ -0,0 +1,31 @@
+use bincode::{Decode, Encode};
```
[the remark this answers is not in the document]
Comment (ryzhyk) [reply adapters/src/test/data.rs:1]: It's under the `test` module.

File: adapters/src/transport/kafka.rs [hunk 2]
```diff
@@ -0,0 +1,647 @@
...
+    }
+
+    /// Validate configuration, set defaults require
+    fn validate(&mut self) -> AnyResult<()> {
+        // Commit automatically.
+        // See https://docs.confluent.io/platform/current/clients/consumer.html#offset-management
+        self.enforce_option("enable.auto.commit", "true")?;
+        self.enforce_option("enable.auto.offset.store", "true")?;
+
+        let group_id = format!(
```
[the remark this answers is not in the document]
Comment (ryzhyk) [reply adapters/src/transport/kafka.rs:158]: yes.

File: adapters/Cargo.toml [hunk 3]
```diff
@@ -16,10 +16,14 @@ once_cell = "1.9.0"
...
+rdkafka = { version = "0.29.0", features = ["cmake-build"] }
 
 [dev-dependencies]
 serde_json = "1.0.87"
 size-of = { version = "0.1.2", features = ["time-std"]}
 tempfile = "3.3.0"
 proptest = "1.0.0"
 proptest-derive = "0.3.0"
+futures = "0.3.25"
+log = "0.4.17"
```
Comment (Kixiron) [inline adapters/Cargo.toml:29]: We should be using [`tracing`](https://docs.rs/tracing/latest/tracing/)

File: adapters/Cargo.toml [hunk 4]
```diff
@@ -16,10 +16,14 @@ once_cell = "1.9.0"
 serde_yaml = "0.9.14"
 csv = { git = "https://github.com/ryzhyk/rust-csv.git" }
 bincode = { version = "2.0.0-rc.2", features = ["serde"] }
+# cmake-build is required on Windows.
+rdkafka = { version = "0.29.0", features = ["cmake-build"] }
```
Comment (Kixiron) [inline adapters/Cargo.toml:20]: Maybe put kafka behind a feature gate

File: adapters/src/transport/kafka.rs [hunk 5]
```diff
@@ -0,0 +1,648 @@
...
+/// consumer gets assigned new partitions, since these new partitions are
+/// may not be in the paused/unpaused state required by the endpoint,
+/// so we may need to pause or unpause them as appropriate.
+///
+/// See https://github.com/edenhill/librdkafka/issues/1849 for a discussion
+/// of the pause/unpause behavior.
+struct KafkaEndpointContext {
+    // We keep a weak reference to the endpoint to avoid a reference cycle:
+    // endpoint->BaseConsumer->context->endpoint.
+    endpoint: Mutex<Weak<KafkaInputEndpointInner>>,
```
Comment (Kixiron) [inline adapters/src/transport/kafka.rs:194]: `parking_lot`

File: adapters/src/transport/kafka.rs [hunk 6]
```diff
@@ -0,0 +1,648 @@
...
+                std::thread::sleep(std::time::Duration::from_millis(1000));
+                // kafka_consumer.poll(POLL_TIMEOUT);
+                // println!("waiting to join the group");
+            } else {
+                break;
+            }
+        }
+
+        let endpoint_clone = endpoint.clone();
+        let _worker = spawn(move || Self::worker_thread(endpoint_clone, consumer));
```
Comment (Kixiron) [inline adapters/src/transport/kafka.rs:287]:
```suggestion
        spawn(move || Self::worker_thread(endpoint_clone, consumer));
```

File: adapters/src/transport/kafka.rs [hunk 7]
```diff
@@ -0,0 +1,648 @@
...
+                        return;
+                    }
+                }
+                Some(Ok(message)) => {
+                    // println!("received {} bytes", message.payload().unwrap().len());
+                    // message.payload().map(|payload| consumer.input(payload));
+
+                    if let Some(payload) = message.payload() {
+                        consumer.input(payload);
+                    };
```
Comment (Kixiron) [inline adapters/src/transport/kafka.rs:376]:
```suggestion
                    }
```

File: adapters/src/transport/kafka.rs [hunk 8]
```diff
@@ -0,0 +1,648 @@
...
+            })
+            .collect::<Vec<_>>();
+        zset_sorted.sort();
+
+        assert_eq!(zset_sorted, data_sorted);
+    }
+
+    proptest! {
+        #![proptest_config(ProptestConfig::with_cases(2))]
+        #[test]
```
Comment (Kixiron) [inline adapters/src/transport/kafka.rs:553]:
```suggestion
        #![proptest_config(ProptestConfig::with_cases(2))]

        #[test]
```

File: adapters/src/transport/kafka.rs [hunk 9]
```diff
@@ -0,0 +1,648 @@
...
+            if matches!(
+                endpoint.kafka_consumer.rebalance_protocol(),
+                RebalanceProtocol::None
+            ) {
+                if attempt == config.group_join_timeout_secs {
+                    Err(AnyError::msg(format!(
+                        "failed to join consumer group '{}', giving up after {}s",
+                        config.kafka_options.get("group.id").unwrap(),
+                        config.group_join_timeout_secs
+                    )))?;
```
Comment (Kixiron) [inline adapters/src/transport/kafka.rs:276]:
```suggestion
                    return Err(AnyError::msg(format!(
                        "failed to join consumer group '{}', giving up after {}s",
                        config.kafka_options.get("group.id").unwrap(),
                        config.group_join_timeout_secs
                    )));
```

File: adapters/src/transport/kafka.rs [hunk 10]
```diff
@@ -0,0 +1,648 @@
...
+    fn post_rebalance(&self, rebalance: &Rebalance<'_>) {
+        // println!("Rebalance: {rebalance:?}");
+        if matches!(rebalance, Rebalance::Assign(_)) {
+            if let Some(endpoint) = self.endpoint.lock().unwrap().upgrade() {
+                if endpoint.state() == PipelineState::Running {
+                    // TODO: handle errors by storing them inside `endpoint`
+                    // for later processing in `poll`.
+                    let _ = endpoint.resume_partitions();
+                } else {
+                    let _ = endpoint.pause_partitions();
```
Comment (Kixiron) [inline adapters/src/transport/kafka.rs:217]: We can also log errors with `tracing::error!()` at the very least

**Review by Kixiron — APPROVED** [review]

**Pull request merged**